### PR TITLE
Add password authentication to Redis ports

### PR DESCRIPTION
**Pull request by pschafhalter**
## What do these changes do?

- Adds password authentication to Redis ports via `ray.init(redis_password="my_password")`
- Throws exception if `redis_password` is set, but Raylet is not running
- Adds a test to verify that Redis ports are secure

## Related issue number

#2925
#2457

### CODE REVIEW

File: python/ray/test/test_ray_init.py
Comment (robertnishihara) [inline python/ray/test/test_ray_init.py]: I *think* the canonical way to write pytest tests is to not use classes and to instead use fixtures, e.g., instead of the `setup`/`teardown`, you can do something like the following.

https://github.com/ray-project/ray/blob/fcdca6de181a7272003a9ddfab27337f4f939373/test/failure_test.py#L29-L35

File: python/ray/services.py
Comment (robertnishihara) [inline python/ray/services.py]: We probably want to remove all instances of the `redis_protected_mode`/`protected_mode` argument, right? And also essentially remove everything from https://github.com/ray-project/ray/pull/2697, right (e.g., the code for making the temp file and stuff)?

File: src/ray/gcs/redis_context.cc
Comment (robertnishihara) [inline src/ray/gcs/redis_context.cc]: This appears in multiple places. Consider moving it to https://github.com/ray-project/ray/blob/master/src/ray/constants.h.

**Comment by pschafhalter** [comment]
Ready for review.
After discussion with @robertnishihara, this PR won't support port authentication for credis. This feature is blocked on review and merge of ray-project/credis#11, so I'll add it to Ray in a follow-up PR.

**Comment by pcmoritz** [comment]
Thanks @pschafhalter, this looks great! Can you add a pointer in the documentation that this exists before we merge it?

**Comment by pschafhalter** [comment]
@pcmoritz where would be a good place to add this in the documentation? The documentation of the Redis password parameters for the Python API and CLI are auto-generated from the code. Do you think it would be good idea to add a seperate security page to the docs?

**Comment by pcmoritz** [comment]
having a separate page "Security" where we describe why this is needed and that it is not a replacement for running stuff behind a firewall and linking to the cluster usage part: https://ray.readthedocs.io/en/latest/using-ray-on-a-cluster.html (and the other cluster setup pages)
would be good. We can also mention in the future that for the autoscaler, security groups are set up automatically.

File: doc/source/security.rst
Comment (robertnishihara) [inline doc/source/security.rst]: typo "unauthorized"

File: python/ray/services.py
Comment (robertnishihara) [reply python/ray/services.py]: A follow up PR is fine.

File: python/ray/test/test_ray_init.py
Comment (robertnishihara) [inline python/ray/test/test_ray_init.py]: Include the full URL

File: python/ray/test/test_ray_init.py
Comment (robertnishihara) [inline python/ray/test/test_ray_init.py]: Why are we calling `ray.shutdown` at the start of the fixture? We don't do this in any of our other tests.

Comment (robertnishihara) [reply python/ray/test/test_ray_init.py]: Instead, our fixtures usually do something like

```python
ray.init()

yield None

ray.shutdown()
```

File: python/ray/test/test_ray_init.py
Comment (robertnishihara) [reply python/ray/test/test_ray_init.py]: Actually, I don't think we need to make a global variable, in generally we should be doing less in the fixture and more in the actual test. Why not move everything to the actual tests and just leave something like 

https://github.com/ray-project/ray/blob/6240ccbc6edeafd2e3223be1edb1448f3ea5d0ca/test/runtest.py#L205-L209

in the fixture?

File: python/ray/test/test_ray_init.py
Comment (robertnishihara) [inline python/ray/test/test_ray_init.py]: What is the purpose of this codepath?

File: python/ray/test/test_ray_init.py
Comment (robertnishihara) [inline python/ray/test/test_ray_init.py]: You can also just use 

https://github.com/ray-project/ray/blob/6240ccbc6edeafd2e3223be1edb1448f3ea5d0ca/test/runtest.py#L2165-L2167

File: python/ray/test/test_ray_init.py
Comment (robertnishihara) [inline python/ray/test/test_ray_init.py]: It's natural to just run the test and if it raises an exception, then pytest will report the error. Instead, we're catching the exception and then asserting that the exception is not `None`, which is very roundabout.

File: python/ray/test/test_ray_init.py
Comment (robertnishihara) [inline python/ray/test/test_ray_init.py]: These tests can probably all be combined into one.

File: src/ray/gcs/redis_context.cc
Comment (robertnishihara) [inline src/ray/gcs/redis_context.cc]: These functions don't follow the naming convention I think. @pcmoritz can comment, but I think it should be `AuthenticateRedis`.

File: python/ray/test/test_ray_init.py
Comment (pschafhalter) [reply python/ray/test/test_ray_init.py]: I intended to test that an exception is thrown when passing a `redis_password` without using raylet, although I can do this instead:
https://github.com/ray-project/ray/blob/6240ccbc6edeafd2e3223be1edb1448f3ea5d0ca/test/runtest.py#L2165-L2167

File: python/ray/test/test_ray_init.py
Comment (robertnishihara) [inline python/ray/test/test_ray_init.py]: Kind of a nit, but `task_id` should be `object_id`.

Also, instead of doing `wait` followed by an assert, maybe just do `ray.get(object_id)`.

Comment (pschafhalter) [reply python/ray/test/test_ray_init.py]: Do you mind if I keep the `ray.wait`? While testing, I ran into a case where `ray.get` wouldn't fail, but block indefinitely (I think this was when I discovered the issues with the new GCS).

**Review by pcmoritz — APPROVED** [review]

**Comment by igoriokas** [comment]
File python/ray/worker.py [line 2084]: redis_password wasn't passed to services.record_log_files_in_redis(...) function.
image

**Comment by pschafhalter** [comment]
@igoriokas thanks! I'll fix this in a follow-up PR.

**Pull request merged**
